Encapsulate upscaling parameters in object [feature/upscale-parameters]

barker29 opened the pull request:
This allows the calling sequences of many constructors/functions to be much shorter and simpler.

barker29 committed:
Begin collecting parameters in UpscaleParameters object.

barker29 committed:
Use UpscaleParameters in FiniteVolumeUpscale class.

barker29 committed:
Use UpscaleParameters object in FiniteVolumeMLMC.

barker29 committed:
Use UpscaleParameters in SpectralAMG_MGL_Coarsener object.

chakshinglee commented on examples/mlmc.cpp:
```diff
@@ -273,19 +273,25 @@ int main(int argc, char* argv[])
 
     // Create Upscaler and Solve
     unique_ptr<FiniteVolumeMLMC> fvupscale;
+    UpscaleParameters param;
```
How do you think about initializing `UpscaleParameters` before command line option parser, and then put the relevant member of `param` (e.g. `param.spect_tol`) directly in `OptionParser`? Something like
```
args.AddOption(&(param.spect_tol), "t", ...
```
This way we can save a few lines initializing each of the parameters in the beginning of `main`, as well as the lines below that assign values. And this comment is relevant for most of the examples/tests files.

barker29 replied:
I like this idea, and I've taken it one step further. Let me know what you think.

chakshinglee replied:
The register function looks good to me (it saves another few lines in each example/test). My only concern would be that new users may be not aware of these command options when looking at the cpp files. Some documentation may help?

barker29 committed:
Read upscaling parameters directly from command line.

barker29 committed:
Add RegisterInOptionsParser() method to UpscalingParameters object
so this object can get parameters from the command line and reduce
duplication in the examples.

barker29 committed:
Use UpscaleParameters.RegisterInOptionsParser() in
finitevolume.cpp, and adjust the use of defaults.

barker29 committed:
UpscaleParameters changes in generalgraph.cpp

barker29 committed:
Switch UpscaleParameters in sampler.cpp and timestep.cpp

chakshinglee commented on examples/graphupscale.cpp:
```diff
@@ -77,8 +83,7 @@ int main(int argc, char* argv[])
         const auto upscale = GraphUpscale(comm, vertex_edge, coarse_factor,
-                                          spect_tol, max_evects, dual_target,
-                                          scaled_dual, energy_dual, hybridization);
+                                          param);
```
very minor issue, this line may join the previous line? Similar issue also for some other files, e.g. `FiniteVolumeMLMC.cpp`.

chakshinglee commented on examples/finitevolume.cpp:
```diff
@@ -73,21 +68,10 @@ int main(int argc, char* argv[])
     bool visualization = false;
     args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                    "--no-visualization", "Enable visualization.");
+    upscale_param.RegisterInOptionsParser(args);
```
I'm thinking adding a comment like
```
// upscaling options from command line, see UpscaleParameters::RegisterInOptionsParser for details
```

barker29 committed:
Move hanging param_ up a line in a few places.

barker29 committed:
Add comment line in example codes to clarify purpose of
RegisterInOptionsParser().

barker29 commented:
Any objections to merging this?

chakshinglee reviewed (approved):
looks good to me
